stl/inc/__msvc_int128.hpp

```diff
@@ -0,0 +1,1155 @@
+        _STL_INTERNAL_CHECK(_Carry < 2);
+#if defined(_M_X64) && !defined(_M_ARM64EC)
+        if (!_STD is_constant_evaluated()) {
+            return _addcarry_u64(_Carry, _Left, _Right, &_Result);
+        }
+#endif // defined(_M_X64) && !defined(_M_ARM64EC)
+
+        const uint64_t _Sum = _Left + _Right + _Carry;
+        _Result             = _Sum;
+        return _Carry ? _Sum <= _Left : _Sum < _Left;
+    }
+
+    static constexpr unsigned char _SubBorrow64(
+        unsigned char _Carry, uint64_t _Left, uint64_t _Right, uint64_t& _Result) noexcept {
+#if defined(_M_X64) && !defined(_M_ARM64EC)
+        if (!_STD is_constant_evaluated()) {
+            return _subborrow_u64(_Carry, _Left, _Right, &_Result);
+        }
+#endif // defined(_M_X64) && !defined(_M_ARM64EC)
+        const auto _Difference = _Left - _Right - _Carry;
+        _Result                = _Difference;
+        return _Carry ? _Difference >= _Left : _Difference > _Left;
+    }
+
+    template <size_t __m, size_t __n>
+    static constexpr void _Knuth_4_3_1_M(
+        const uint32_t (&__u)[__m], const uint32_t (&__v)[__n], uint32_t (&__w)[__n + __m]) noexcept {
+        _STL_INTERNAL_STATIC_ASSERT(__m <= numeric_limits<int>::max());
+        _STL_INTERNAL_STATIC_ASSERT(__n <= numeric_limits<int>::max());
+
+        for (auto& _Elem : __w) {
+            _Elem = 0;
+        }
+
+        for (int __j = 0; __j < static_cast<int>(__n); ++__j) {
+            // stash Knuth's `k` in the lower 32 bits of __t
+            uint64_t __t = 0;
+            for (int __i = 0; __i < static_cast<int>(__m); ++__i) {
+                __t += static_cast<uint64_t>(__u[__i]) * __v[__j] + __w[__i + __j];
+                __w[__i + __j] = static_cast<uint32_t>(__t);
+                __t >>= 32;
+            }
+            __w[__j + __m] = static_cast<uint32_t>(__t);
+        }
+    }
+
+    _NODISCARD static constexpr uint64_t _UMul128(
+        const uint64_t _Left, const uint64_t _Right, uint64_t& _High_result) noexcept {
+#if defined(_M_X64) && !defined(_M_ARM64EC)
+        if (!_STD is_constant_evaluated()) {
+            return _umul128(_Left, _Right, &_High_result);
+        }
+#endif // defined(_M_X64) && !defined(_M_ARM64EC)
+
+        const uint32_t __u[2] = {
+            static_cast<uint32_t>(_Left),
+            static_cast<uint32_t>(_Left >> 32),
+        };
+        const uint32_t __v[2] = {
+            static_cast<uint32_t>(_Right),
+            static_cast<uint32_t>(_Right >> 32),
+        };
+        uint32_t __w[4];
+
+        // multiply 2-digit numbers with 4-digit result in base 2^32
+        _Knuth_4_3_1_M(__u, __v, __w);
+
+        _High_result = (static_cast<uint64_t>(__w[3]) << 32) | __w[2];
+        return (static_cast<uint64_t>(__w[1]) << 32) | __w[0];
+    }
+
+    static constexpr void _Knuth_4_3_1_D_impl(uint32_t __u[/* __n + __m + 1 */], const uint32_t __v[/* __n */],

```

Comment:
This function intentionally uses pointers and counts so three different callers get the same function instead of instantiating all of this three times. There's little to gain from loop unrolling, except for the special case when there's only a single quotient digit and we don't care about the remainder, which none of the callers uses. (The handrolled 2-digit base-2^64 flavor in `operator/(_Uint128, _Uint128)` is that special case, but doesn't call this.)

Reply:
(Resolving this without prejudice, feel free to reopen.)